File: app/src/main/java/com/example/sleppify/TaskReminderPrepareWorker.java
package com.example.sleppify;

import android.content.Context;
import android.content.SharedPreferences;
import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.work.Data;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicReference;

public class TaskReminderPrepareWorker extends Worker {

    public TaskReminderPrepareWorker(@NonNull Context context, @NonNull WorkerParameters workerParams) {
        super(context, workerParams);
    }

    @NonNull
    @Override
    public Result doWork() {
        Context context = getApplicationContext();
        SharedPreferences settingsPrefs = context.getSharedPreferences(CloudSyncManager.PREFS_SETTINGS, Context.MODE_PRIVATE);
        boolean smartSuggestionsEnabled = settingsPrefs.getBoolean(
                CloudSyncManager.KEY_SMART_SUGGESTIONS_ENABLED,
                settingsPrefs.getBoolean(CloudSyncManager.KEY_AI_SHIFT_ENABLED, true)
        );
        if (!smartSuggestionsEnabled) {
            return Result.success();
        }

        Data input = getInputData();
        String reminderId = input.getString(TaskReminderConstants.INPUT_REMINDER_ID);
        String taskTitle = input.getString(TaskReminderConstants.INPUT_TASK_TITLE);
        String taskDesc = input.getString(TaskReminderConstants.INPUT_TASK_DESC);
        String taskTime = input.getString(TaskReminderConstants.INPUT_TASK_TIME);
        String dateKey = input.getString(TaskReminderConstants.INPUT_DATE_KEY);
        long notifyAtMs = input.getLong(TaskReminderConstants.INPUT_NOTIFY_AT_MS, 0L);

        if (TextUtils.isEmpty(reminderId) || TextUtils.isEmpty(taskTitle)) {
            return Result.success();
        }

        String profileName = resolveProfileName(context);
        String snapshot = buildTaskSnapshot(taskTitle, taskDesc, taskTime, dateKey);

        AtomicReference<String> messageRef = new AtomicReference<>(null);
        CountDownLatch latch = new CountDownLatch(1);

        new GeminiIntelligenceService().generateTodayAgendaSummary(
                profileName,
                snapshot,
                new GeminiIntelligenceService.TodayAgendaSummaryCallback() {
                    @Override
                    public void onSuccess(String message) {
                        messageRef.set(message);
                        latch.countDown();
                    }

                    @Override
                    public void onError(String error) {
                        latch.countDown();
                    }
                }
        );

        boolean completed;
        try {
            completed = latch.await(20, TimeUnit.SECONDS);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            return Result.retry();
        }

        String summary = messageRef.get();
        if (!completed || TextUtils.isEmpty(summary)) {
            long nowMs = System.currentTimeMillis();
            if (notifyAtMs > nowMs + 15_000L) {
                return Result.retry();
            }
            summary = buildFallbackSummary(taskTitle, taskDesc, taskTime);
        }

        SharedPreferences cachePrefs = context.getSharedPreferences(TaskReminderConstants.PREFS_CACHE, Context.MODE_PRIVATE);
        cachePrefs.edit()
                .putString(TaskReminderConstants.KEY_SUMMARY_PREFIX + reminderId, summary)
                .putLong(TaskReminderConstants.KEY_SUMMARY_TS_PREFIX + reminderId, System.currentTimeMillis())
                .apply();

        return Result.success();
    }

    @NonNull
    private String resolveProfileName(@NonNull Context context) {
        try {
            AuthManager authManager = AuthManager.getInstance(context);
            if (authManager != null && authManager.isSignedIn()) {
                String displayName = authManager.getDisplayName();
                if (!TextUtils.isEmpty(displayName)) {
                    return displayName;
                }
            }
        } catch (Throwable ignored) {
            // Sin sesion valida en background usamos valor neutro.
        }
        return "usuario";
    }

    @NonNull
    private String buildTaskSnapshot(
            @NonNull String taskTitle,
            @Nullable String taskDesc,
            @Nullable String taskTime,
            @Nullable String dateKey
    ) {
        String safeDesc = taskDesc == null ? "" : taskDesc.trim();
        String safeTime = taskTime == null ? "" : taskTime.trim();
        String safeDate = dateKey == null ? "" : dateKey.trim();

        StringBuilder builder = new StringBuilder();
        builder.append("total_tareas_hoy=1; ");
        builder.append("lista=[1) ").append(taskTitle);
        if (!safeDesc.isEmpty()) {
            builder.append(" - ").append(safeDesc);
        }
        if (!safeTime.isEmpty()) {
            builder.append(" - hora ").append(safeTime);
        }
        if (!safeDate.isEmpty()) {
            builder.append(" - fecha ").append(safeDate);
        }
        builder.append("]");
        return builder.toString();
    }

    @NonNull
    private String buildFallbackSummary(
            @NonNull String taskTitle,
            @Nullable String taskDesc,
            @Nullable String taskTime
    ) {
        String safeDesc = taskDesc == null ? "" : taskDesc.trim();
        String safeTime = taskTime == null ? "" : taskTime.trim();

        if (!safeDesc.isEmpty() && !safeTime.isEmpty()) {
            return "Recordatorio breve: " + taskTitle + " a las " + safeTime + ". " + safeDesc;
        }
        if (!safeTime.isEmpty()) {
            return "Recordatorio breve: " + taskTitle + " a las " + safeTime + ".";
        }
        if (!safeDesc.isEmpty()) {
            return "Recordatorio breve: " + taskTitle + ". " + safeDesc;
        }
        return "Recordatorio breve: toca avanzar en \"" + taskTitle + "\".";
    }
}

File: app/src/main/java/com/example/sleppify/TaskReminderPrepareWorker.kt
package com.example.sleppify

import android.content.Context
import android.text.TextUtils
import androidx.work.Data
import androidx.work.Worker
import androidx.work.WorkerParameters

class TaskReminderPrepareWorker(context: Context, workerParams: WorkerParameters) : Worker(context, workerParams) {

    override fun doWork(): Result {
        val context = applicationContext
        val settingsPrefs = context.getSharedPreferences(CloudSyncManager.PREFS_SETTINGS, Context.MODE_PRIVATE)
        val smartSuggestionsEnabled = settingsPrefs.getBoolean(
            CloudSyncManager.KEY_SMART_SUGGESTIONS_ENABLED,
            settingsPrefs.getBoolean(CloudSyncManager.KEY_AI_SHIFT_ENABLED, true)
        )
        if (!smartSuggestionsEnabled) {
            return Result.success()
        }

        val input: Data = inputData
        val reminderId = input.getString(TaskReminderConstants.INPUT_REMINDER_ID)
        val taskTitle = input.getString(TaskReminderConstants.INPUT_TASK_TITLE)
        val taskDesc = input.getString(TaskReminderConstants.INPUT_TASK_DESC)
        val taskTime = input.getString(TaskReminderConstants.INPUT_TASK_TIME)
        val dateKey = input.getString(TaskReminderConstants.INPUT_DATE_KEY)
        val notifyAtMs = input.getLong(TaskReminderConstants.INPUT_NOTIFY_AT_MS, 0L)
        val cachePrefs = context.getSharedPreferences(TaskReminderConstants.PREFS_CACHE, Context.MODE_PRIVATE)

        if (TextUtils.isEmpty(reminderId) || TextUtils.isEmpty(taskTitle)) {
            return Result.success()
        }

        // AI is intentionally NOT invoked from background workers. Per product rule,
        // the only allowed Gemini triggers are pull-to-refresh and task creation.
        // This worker now always builds a local fallback summary.
        val summary = buildFallbackSummary(taskTitle!!, taskDesc, taskTime)

        cachePrefs.edit()
            .putString(TaskReminderConstants.KEY_SUMMARY_PREFIX + reminderId, summary)
            .putLong(TaskReminderConstants.KEY_SUMMARY_TS_PREFIX + reminderId, System.currentTimeMillis())
            .apply()

        return Result.success()
    }

    private fun resolveProfileName(context: Context): String {
        return try {
            val authManager = AuthManager.getInstance(context)
            if (authManager.isSignedIn()) {
                val displayName = authManager.getDisplayName()
                if (!displayName.isNullOrBlank()) {
                    return displayName
                }
            }
            "usuario"
        } catch (_: Throwable) {
            "usuario"
        }
    }

    private fun buildTaskSnapshot(
        taskTitle: String,
        taskDesc: String?,
        taskTime: String?,
        dateKey: String?
    ): String {
        val safeDesc = taskDesc?.trim().orEmpty()
        val safeTime = taskTime?.trim().orEmpty()
        val safeDate = dateKey?.trim().orEmpty()

        return buildString {
            append("total_tareas_hoy=1; ")
            append("lista=[1) ")
            append(taskTitle)
            if (safeDesc.isNotEmpty()) {
                append(" - ")
                append(safeDesc)
            }
            if (safeTime.isNotEmpty()) {
                append(" - hora ")
                append(safeTime)
            }
            if (safeDate.isNotEmpty()) {
                append(" - fecha ")
                append(safeDate)
            }
            append("]")
        }
    }

    private fun buildFallbackSummary(taskTitle: String, taskDesc: String?, taskTime: String?): String {
        val safeDesc = taskDesc?.trim().orEmpty()
        val safeTime = taskTime?.trim().orEmpty()

        return when {
            safeDesc.isNotEmpty() && safeTime.isNotEmpty() -> "Recordatorio breve: $taskTitle a las $safeTime. $safeDesc"
            safeTime.isNotEmpty() -> "Recordatorio breve: $taskTitle a las $safeTime."
            safeDesc.isNotEmpty() -> "Recordatorio breve: $taskTitle. $safeDesc"
            else -> "Recordatorio breve: toca avanzar en \"$taskTitle\"."
        }
    }
}
